### PR TITLE
Add pipeline permissions

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-506

#### Link to staging deployment URL 

N.A.

#### Links to any Pull Requests related to this

https://github.com/biomage-ltd/api/pull/58

#### Anything else the reviewers should know about the changes here

- Added permissions to API to function with pipelines

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by ivababukova** [comment]
have you tested it in aws to see if it works?

**Review by marcellp — CHANGES_REQUESTED** [review]
Can we add `DescribeExecution` and `DescribeStateMachine` as well here? We will be using those to check if a pipeline is currently running so we can disable parts of the website accordingly.

File: cf/irsa-api-role.yaml [hunk 1]
```diff
@@ -81,22 +80,42 @@ Resources:
...
                   - !Sub "arn:aws:eks:${AWS::Region}:${AWS::AccountId}:cluster/biomage-${Environment}"
+        - PolicyName: !Sub "can-manage-state-machine-${Environment}"
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - "states:CreateStateMachine"
+                  - "states:UpdateStateMachine"
+                  - "states:StartExecution"
```
Comment (marcellp) [inline cf/irsa-api-role.yaml:111]: This should probably fall under the `can-manage-execution-${Environment}` policy.

Comment (aerlaut) [reply cf/irsa-api-role.yaml:111]: Oh ok, so I just put the two resource types (statement and execution) under one policy?

Comment (aerlaut) [reply cf/irsa-api-role.yaml:111]: updated in latest commit. Can you please take a look at it again?

**Comment by aerlaut** [comment]
> have you tested it in aws to see if it works?

This script should add permissions to a service right? I don't really know how to test it. Do I upload through cloudformation in the web console?

I tried validating this file using `aws cloudformation validate-template` but I don't have permission.

I have not yet tested it functionally. For the pipeline to run end to end,  it still need https://github.com/biomage-ltd/api/pull/58. Marcell and I have tested the both locally and it's working end to end.

**Comment by ivababukova** [comment]
> > have you tested it in aws to see if it works?
> 
> This script should add permissions to a service right? I don't really know how to test it. Do I upload through cloudformation in the web console?
> 
> I tried validating this file using `aws cloudformation validate-template` but I don't have permission.
> 
> I have not yet tested it functionally. For the pipeline to run end to end, it still need [biomage-ltd/api#58](https://github.com/biomage-ltd/api/pull/58). Marcell and I have tested the both locally and it's working end to end.

@marcellp how do you test the changes of roles? Last time I did cloudformation changes, I would test them by uploading the cloudformation file in aws cloudformation console and verifying that it does not fail validation. You did a lot of changes since last time I touched that repo, so I am not sure now whether there isn't any automated way to test. Also, after this gets merged, do we deploy it to aws via github actions or do we need to update the role with the new version of the cloudformation manually?

**Comment by marcellp** [comment]
No, everything works automatically @ivababukova -- there is a pipeline here (you canse under `3 successful checks`) that tries to lint the CF file and do a dry run to make sure it works. Then, when it is merged, it is automatically deployed.

**Comment by ivababukova** [comment]
In that case Agi, no need to test it in any other way apart from merging and hoping for the best

**Review by ivababukova — APPROVED** [review]

File: cf/irsa-api-role.yaml [hunk 2]
```diff
@@ -81,22 +80,37 @@ Resources:
...
+                Action:
+                  - "states:DescribeStateMachine"
+                  - "states:CreateStateMachine"
+                  - "states:UpdateStateMachine"
+                  - "states:DescribeExecution"
+                  - "states:StartExecution"
+                  - "states:StopExecution"
+                Resource:
+                  - !Sub "arn:aws:states:${AWS::Region}:${AWS::AccountId}:stateMachine:biomage-pipeline-${Environment}-*"
+                  - !Sub "arn:aws:states:${AWS::Region}:${AWS::AccountId}:execution:biomage-pipeline-${Environment}-*:*"
```
Comment (marcellp) [inline cf/irsa-api-role.yaml]: Even though this is supposedly valid, I think we should break up the state machine roles from the execution roles, and have them as separate policies.

That is, there should be a policy with `states:DescribeStateMachine` `states:CreateStateMachine` and `states:UpdateStateMachine` with the resource `!Sub "arn:aws:states:${AWS::Region}:${AWS::AccountId}:stateMachine:biomage-pipeline-${Environment}-*"`, and a separate one for the executions accordingly.

When you create a role, it's usually a good idea to create different policies depending on the nature of the resource. That is, one policy per principal (state machines and executions, in this case), with actions that can only be performed on the appropriate resources in the policy.

Other than that, lgtm.

**Review by marcellp — CHANGES_REQUESTED** [review]

File: cf/irsa-api-role.yaml (hunk 2 above)
Comment (aerlaut) [reply cf/irsa-api-role.yaml]: According the [AWS documentation](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsstepfunctions.html#awsstepfunctions-activity), the following permissions are available to state machines : 
 - `states:DescribeStateMachine`
  - `states:CreateStateMachine`
  - `states:UpdateStateMachine`
  - `states:StartExecution`

While these are available to executions
  - `states:DescribeExecution`
  - `states:StopExecution`
  
It's a bit weird to have `StartExecution` under `StateMachine`, but it makes sense

**Comment by marcellp** [comment]
This is correct per the permissions document [here](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsstepfunctions.html#awsstepfunctions-activity).

Approved.

**Pull request merged**